### PR TITLE
SAS chunksize / iteration issues

### LINKED ISSUE #13654: Issue iterating with pd.read_sas
#### Code Sample, a copy-pastable example if possible

```
import pandas as pd

f = pd.read_sas('test.sas7bdat', iterator=True)
for chunk in f:
    print(f)

---------------------------------------------------------------------------
AbstractMethodError                       Traceback (most recent call last)
<ipython-input-3-bedf2769bffe> in <module>()
----> 1 for chunk in f:
      2     print(chunk)

C:\Anaconda3\lib\site-packages\pandas\io\common.py in __next__(self)
     99 
    100     def __next__(self):
--> 101         raise AbstractMethodError(self)
    102 
    103 if not compat.PY3:

AbstractMethodError: This method must be defined in the concrete class of SAS7BDATReader
```
#### Expected Output

Print of each chunk

I'm no expert, but it seems as if the SAS7BDAT class inherits from BaseIterator but does not provide its own `__next__` method
## INSTALLED VERSIONS

commit: None
python: 3.5.1.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None

pandas: 0.18.1
nose: 1.3.7
pip: 8.1.2
setuptools: 20.3
Cython: 0.23.4
numpy: 1.10.4
scipy: 0.17.0
statsmodels: 0.6.1
xarray: None
IPython: 4.1.2
sphinx: 1.3.1
patsy: 0.4.1
dateutil: 2.5.1
pytz: 2016.2
blosc: None
bottleneck: 1.0.0
tables: 3.2.2
numexpr: 2.5.2
matplotlib: 1.5.1
openpyxl: 2.3.2
xlrd: 0.9.4
xlwt: 1.0.0
xlsxwriter: 0.8.4
lxml: 3.6.0
bs4: 4.4.1
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.12
pymysql: None
psycopg2: None
jinja2: 2.8
boto: 2.39.0
pandas_datareader: None

### LINKED ISSUE #14734: read_sas with chunksize/iterator raises ValueError
`read_sas` doesn't work well with `chunksize` or `iterator` parameters.

#### Code Sample and Problem Description
The following data test file in the repository have 32 lines.
```python
sasfile = 'pandas/io/tests/sas/data/airline.sas7bdat'
pd.read_sas(sasfile).shape
Out[18]: (32, 6)
```
When we carefully read the file with `chunksize`/`iterator`, all's well:
```python
reader = pd.read_sas(sasfile, chunksize=16)
df = reader.read()
df.shape
Out[31]: (16, 6)
df = reader.read()
df.shape
Out[33]: (16, 6)
```
or 
```python
reader = pd.read_sas(sasfile, iterator=True)
df = reader.read(30)
df.shape
Out[37]: (30, 6)
df = reader.read(2)
df.shape
Out[39]: (2, 6)
df = reader.read(2)
type(df)
Out[41]: NoneType
```
But if we don't know the length of the data, we'll easily stumble on an exception and won't read the whole data, which is painful with large files.
```python
reader = pd.read_sas(sasfile, chunksize=20)
df = reader.read()
df.shape
Out[45]: (20, 6)
df = reader.read()
Traceback (most recent call last):
  File "/usr/local/lib64/python3.5/site-packages/IPython/core/interactiveshell.py", line 2885, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-46-c5d811b93ac1>", line 1, in <module>
    df = reader.read()
  File "/home/users/piotr/workspace/pandas-pijucha/pandas/io/sas/sas7bdat.py", line 604, in read
    rslt = self._chunk_to_dataframe()
  File "/home/users/piotr/workspace/pandas-pijucha/pandas/io/sas/sas7bdat.py", line 646, in _chunk_to_dataframe
    dtype=self.byte_order + 'd')
  File "/home/users/piotr/workspace/pandas-pijucha/pandas/core/frame.py", line 2419, in __setitem__
    self._set_item(key, value)
  File "/home/users/piotr/workspace/pandas-pijucha/pandas/core/frame.py", line 2485, in _set_item
    value = self._sanitize_column(key, value)
  File "/home/users/piotr/workspace/pandas-pijucha/pandas/core/frame.py", line 2656, in _sanitize_column
    value = _sanitize_index(value, self.index, copy=False)
  File "/home/users/piotr/workspace/pandas-pijucha/pandas/core/series.py", line 2793, in _sanitize_index
    raise ValueError('Length of values does not match length of ' 'index')
ValueError: Length of values does not match length of index
```
or
```python
reader = pd.read_sas(sasfile, iterator=True)
reader.read(30).shape
Out[51]: (30, 6)
reader.read(30).shape
Traceback (most recent call last):
  File "/usr/local/lib64/python3.5/site-packages/IPython/core/interactiveshell.py", line 2885, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-52-5d757f713808>", line 1, in <module>
    reader.read(30).shape
  File "/home/users/piotr/workspace/pandas-pijucha/pandas/io/sas/sas7bdat.py", line 604, in read
    rslt = self._chunk_to_dataframe()
  File "/home/users/piotr/workspace/pandas-pijucha/pandas/io/sas/sas7bdat.py", line 646, in _chunk_to_dataframe
    dtype=self.byte_order + 'd')
  File "/home/users/piotr/workspace/pandas-pijucha/pandas/core/frame.py", line 2419, in __setitem__
    self._set_item(key, value)
  File "/home/users/piotr/workspace/pandas-pijucha/pandas/core/frame.py", line 2485, in _set_item
    value = self._sanitize_column(key, value)
  File "/home/users/piotr/workspace/pandas-pijucha/pandas/core/frame.py", line 2656, in _sanitize_column
    value = _sanitize_index(value, self.index, copy=False)
  File "/home/users/piotr/workspace/pandas-pijucha/pandas/core/series.py", line 2793, in _sanitize_index
    raise ValueError('Length of values does not match length of ' 'index')
ValueError: Length of values does not match length of index
```
#### Output of ``pd.show_versions()``

<details>
pd.show_versions()

INSTALLED VERSIONS
------------------
commit: 75b606abad51762b3faace1bbfa0a8d2241dc297
python: 3.5.1.final.0
python-bits: 64
OS: Linux
OS-release: 4.1.20-1
machine: x86_64
processor: Intel(R)_Core(TM)_i5-2520M_CPU_@_2.50GHz
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.19.0+112.g75b606a
nose: 1.3.7
pip: 9.0.1
setuptools: 21.0.0
Cython: 0.24
numpy: 1.11.0
</details>

### PR DESCRIPTION
**Pull request by kshedden**
closes #14734
closes #13654
 - [X] tests added / passed
 - [X] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

File: pandas/io/tests/sas/test_sas7bdat.py [hunk 1]
```diff
@@ -65,6 +65,32 @@ def test_from_iterator(self):
                 df = rdr.read(3)
                 tm.assert_frame_equal(df, df0.iloc[2:5, :])
 
+    def test_iterator_loop(self):
+        for j in 0, 1:
```
Comment (jreback) [inline pandas/io/tests/sas/test_sas7bdat.py]: can you add the issue references here

File: pandas/io/tests/sas/test_sas7bdat.py [hunk 2]
```diff
@@ -65,6 +65,32 @@ def test_from_iterator(self):
...
+                    fname = os.path.join(self.dirpath, "test%d.sas7bdat" % k)
+                    with open(fname, 'rb') as f:
+                        byts = f.read()
+                    buf = io.BytesIO(byts)
+                    rdr = pd.read_sas(buf, format="sas7bdat",
+                                      chunksize=chunksize, encoding='utf-8')
+                    y = 0
+                    for x in rdr:
+                        y += x.shape[0]
+                    assert(y == rdr.row_count)
```
Comment (jreback) [inline pandas/io/tests/sas/test_sas7bdat.py]: self.assertTrue

**Comment by jreback** [comment]
pls add to the whatsnew. ping when ready / green.

**Comment by jorisvandenbossche** [comment]
Question: this is not a problem in the xport reader? If so, is there a test that confirms this?

**Comment by kshedden** [comment]
It works there but I added a few tests.

**Comment by jorisvandenbossche** [comment]
@kshedden Thanks. You can move the whatsnew notice to the 0.19.2 file

File: pandas/io/tests/sas/test_sas7bdat.py [hunk 3]
```diff
@@ -65,6 +65,33 @@ def test_from_iterator(self):
...
 
+    def test_iterator_loop(self):
+        # github #13654
+        for j in 0, 1:
+            for k in self.test_ix[j]:
+                for chunksize in 3, 5, 10, 11:
+                    fname = os.path.join(self.dirpath, "test%d.sas7bdat" % k)
+                    with open(fname, 'rb') as f:
+                        byts = f.read()
+                    buf = io.BytesIO(byts)
```
Comment (jorisvandenbossche) [inline pandas/io/tests/sas/test_sas7bdat.py]: Is there a reason you read the file into a bytes object and not just pass the `fname` to ``read_sas`` ?

Comment (kshedden) [inline pandas/io/tests/sas/test_sas7bdat.py]: It seems to work either way so I took all but one of the bytesio tests and simplified to a direct file read.

File: pandas/io/tests/sas/test_sas7bdat.py [hunk 4]
```diff
@@ -44,27 +44,46 @@ def test_from_buffer(self):
             df0 = self.data[j]
             for k in self.test_ix[j]:
                 fname = os.path.join(self.dirpath, "test%d.sas7bdat" % k)
-                with open(fname, 'rb') as f:
-                    byts = f.read()
-                buf = io.BytesIO(byts)
-                df = pd.read_sas(buf, format="sas7bdat", encoding='utf-8')
```
Comment (jorisvandenbossche) [inline pandas/io/tests/sas/test_sas7bdat.py]: I think it is more logical to keep the buffer reading here, as this test is called "test_from_buffer" is I suppose is exactly testing this (and  then maybe use plain reading from file in "test_iterator_read_too_much")

**Comment by jorisvandenbossche** [comment]
@kshedden Thanks for the quick fix!

**Pull request merged**